swarm/network/stream/syncer.go

```diff
@@ -69,76 +68,103 @@ func (s *SwarmSyncerServer) Close() {
 
 // GetData retrieves the actual chunk from netstore
 func (s *SwarmSyncerServer) GetData(ctx context.Context, key []byte) ([]byte, error) {
-	chunk, err := s.store.Get(ctx, storage.Address(key))
+	ch, err := s.store.Get(ctx, chunk.ModeGetSync, storage.Address(key))
 	if err != nil {
 		return nil, err
 	}
-	return chunk.Data(), nil
+	return ch.Data(), nil
 }
 
 // SessionIndex returns current storage bin (po) index.
 func (s *SwarmSyncerServer) SessionIndex() (uint64, error) {
-	return s.store.BinIndex(s.po), nil
+	return s.store.LastPullSubscriptionBinID(s.po)
 }
 
-// GetBatch retrieves the next batch of hashes from the dbstore
+// SetNextBatch retrieves the next batch of hashes from the localstore.
+// It expects a range if bin IDs, both ends inclusive in syncing, and returns
+// concatenated byte slice of chunk addresses and bin IDs of the first and
+// the last one in that slice. The batch may have up to BatchSize number of
+// chunk addresses. If at least one chunk is added to the batch and no new chunks
+// are added in batchTimeout period, the batch will be returned. This function
+// will block until new chunks are received from localstore pull subscription.
 func (s *SwarmSyncerServer) SetNextBatch(from, to uint64) ([]byte, uint64, uint64, *HandoverProof, error) {
-	var batch []byte
-	i := 0
-
-	var ticker *time.Ticker
+	descriptors, stop := s.store.SubscribePull(context.Background(), s.po, from, to)
+	defer stop()
+
+	const batchTimeout = 2 * time.Second
+
+	var (
+		batch        []byte
+		batchSize    int
+		batchStartID *uint64
+		batchEndID   uint64
+		timer        *time.Timer
+		timerC       <-chan time.Time
+	)
 	defer func() {
-		if ticker != nil {
-			ticker.Stop()
+		if timer != nil {
+			timer.Stop()
 		}
 	}()
-	var wait bool
-	for {
-		if wait {
-			if ticker == nil {
-				ticker = time.NewTicker(1000 * time.Millisecond)
+
+	for iterate := true; iterate; {
+		select {
+		case d, ok := <-descriptors:
+			if !ok {
+				iterate = false
+				break
 			}
-			select {
-			case <-ticker.C:
-			case <-s.quit:
-				return nil, 0, 0, nil, nil
+			batch = append(batch, d.Address[:]...)
+			// This is the most naive approach to label the chunk as synced
+			// allowing it to be garbage collected. A proper way requires
+			// validating that the chunk is successfully stored by the peer.
+			err := s.store.Set(context.Background(), chunk.ModeSetSync, d.Address)
```

Comment:
@zelig I added set sync call here opposed to change localstore.DB.Put to always add chunk to the garbage collection index. Adding that functionality beaks a lot of localstore tests that explicitly check that the chunk is not garbage collected until set sync is called in various scenarios.

In case that Put change is a preferred way, this is the required change:

```diff
diff --git a/swarm/storage/localstore/mode_put.go b/swarm/storage/localstore/mode_put.go
index ecad86dac..d6e982d62 100644
--- a/swarm/storage/localstore/mode_put.go
+++ b/swarm/storage/localstore/mode_put.go
@@ -147,6 +147,29 @@ func (db *DB) put(mode chunk.ModePut, item shed.Item) (err error) {
                return ErrInvalidMode
        }
 
+       // WARNING: This is a temporary approach to insert
+       // the chunk in GC index for all Put modes.
+       // The code under this warning, should be removed
+       // when SetModeSync is implemented in stream package.
+       if mode != chunk.ModePutRequest {
+               i, err := db.retrievalAccessIndex.Get(item)
+               switch err {
+               case nil:
+                       item.AccessTimestamp = i.AccessTimestamp
+                       db.gcIndex.DeleteInBatch(batch, item)
+                       gcSizeChange--
+               case leveldb.ErrNotFound:
+                       // the chunk is not accessed before
+               default:
+                       return err
+               }
+               item.AccessTimestamp = now()
+               db.retrievalAccessIndex.PutInBatch(batch, item)
+               db.gcIndex.PutInBatch(batch, item)
+               gcSizeChange++
+       }
+       // END WARNING
+
        err = db.incGCSizeInBatch(batch, gcSizeChange)
        if err != nil {
                return err
```

But localstore tests need to be changed and commented how to revert them later.